LUCENE-8778: Define analyzer SPI names as static final fields and document the names

mocobeta opened the pull request:
See: https://issues.apache.org/jira/browse/LUCENE-8778

Changes in this PR:
- Define SPI names as static final fields.
- Document SPI names by custom Javadoc tag `@lucene.spi`.
- Add validation rules for `NAME` field definition and the Javadoc tag.
- Change SPI loader to get the names from the new `NAME` fields

mocobeta commented:
To begin with, I only changed char filters for design review.

uschindler commented on lucene/analysis/common/src/java/org/apache/lucene/analysis/util/AnalysisSPILoader.java:
```diff
@@ -83,17 +74,22 @@ public synchronized void reload(ClassLoader classloader) {
+        // Lookup "NAME" field with appropriate modifiers. Also it must be a String field.
+        final Field field = service.getField("NAME");
+        int modifier = field.getModifiers();
+        if (Modifier.isPublic(modifier) && Modifier.isStatic(modifier) && Modifier.isFinal(modifier) &&
```
Actually, you can leave the check inside to make it more explicit, I just wanted to mention this: <https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Class.html#getField(java.lang.String)>: "Returns a Field object that reflects the specified public member field of the class or interface represented by this Class object"

mocobeta replied:
Thanks for the comment,
I noticed another method `Class#getDeclaredField(String)` which does not traverse superclasses.
https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Class.html#getDeclaredField(java.lang.String)
Should we use this instead of `getField(String)` ?

uschindler replied:
I was thinking last night about this, too! My answer: YES! This allows us to have a full check that everything is correct.

uschindler replied:
But I have no idea about the consequences with security manager... So maybe check this.

uschindler replied:
I have a better idea: Use `getField()` and remove the `isPublic()` check. But do `Objects.equals(field.getDeclaringClass(),  service)` as safety check. That has no security manager problems and does the check, too.

uschindler replied:
The general problem of the `getDeclaredField()` for SPI is: The SPI interfaces may be loaded from another classloader (e.g., in Solr) - becaus ethat's the whole idea about SPI!

mocobeta replied:
OK, I will change the safety check in that manner.

mocobeta commented:
I updated the branch: refine the catch clause and safety checks in the SPI loader; use camel casing for names.

Remaining work:
- Add the "NAME" field and Javadocs to all other analysis components
- Write regression tests for all existing components
- (Look for easier ways to do so...)

mocobeta commented:
"NAME" field and Javadoc tags were added to all tokenizers/filters so this passes unit tests. I will do regression test soon.

uschindler commented:
This looks good. Was a good idea to have a converter program.

mocobeta commented:
It is not related to this issue, but I'd like to add this line to the gitignore file. I occasionally see the python bytecode cache directory is created under `dev-tools/scripts`, though I've not figured out which ant task generates this... 
```
diff --git a/.gitignore b/.gitignore
index 4b947436dc..674ddd198f 100644
--- a/.gitignore
+++ b/.gitignore
@@ -26,3 +26,4 @@ pom.xml
 /nbproject
 /nb-build
 .pydevproject
+__pycache__
```

Alternatively, I think we can completely disable the cache (don't write .pyc or .pyo files) by `-B` option.
https://stackoverflow.com/questions/16869024/what-is-pycache

mocobeta commented:
Just a quick note:
A few times I nearly committed the `__pycache__` directory, so I added this to the gitignore file.
This does not work for python2. Please let me know if the python scripts could be run on python2 (depending on the system environments), I fixed the file again.

mocobeta commented:
Hi @uschindler,

this PR passed all tests & precommit. If the changes in `AnalysisSPILoader` looks fine for you, I will commit it to the ASF repo's master branch. (Maybe I will merge it to the master, but not to 8x?) Let me know if you find anything wrong, or of course you can directly modify the branch like the previous time.

mocobeta commented:
Hi,
just from curiosity I've googled for a while if we can handle case insensitive lookup with LinkedHashMap so that we don't need extra Set object for reference. 

There are (not very attractive) two options: 
1. Create a special Map implementation which preserves insert order and also provides case insensitive lookup
2. Create a special class which holds a String (service name) and has case-insensitive hashCode() and equals() 

I think both are somewhat over-engineering for the use case here, so current implementation seems like an appropriate solution. Any comments?

uschindler commented:
Hi,
I am fine with all changes. +1 to commit. Sorry for the delay, I though it was already done!
I would not change the map implementation, it would be over-engineered. The additional heap space would also be there with an alternate implementation.

mocobeta commented:
Thanks, I will push it to master soon.

mocobeta commented:
Merged into upstream master branch.

Updated CHANGES:
```
+* LUCENE-8778: Define analyzer SPI names as static final fields and document the names in Javadocs.
+  (Tomoko Uchida, Uwe Schindler)
```
